File: web/src/types/forecast.ts
export const WEATHER_MODELS = [
  'icon_d2',
  'icon_eu',
  'icon_global',
  'ecmwf_ifs025',
  'ecmwf_aifs025',
  'gfs_seamless',
  'meteofrance_arome_france_hd',
  'meteofrance_arpege_europe',
  'jma_msm',
  'gem_seamless',
] as const

export type WeatherModelId = (typeof WEATHER_MODELS)[number]

export const MODEL_DISPLAY_NAMES: Record<WeatherModelId, string> = {
  icon_d2: 'ICON-D2 · DWD',
  icon_eu: 'ICON-EU · DWD',
  icon_global: 'ICON · DWD',
  ecmwf_ifs025: 'ECMWF IFS',
  ecmwf_aifs025: 'ECMWF AIFS',
  gfs_seamless: 'GFS · NOAA',
  meteofrance_arome_france_hd: 'AROME · MeteoFrance',
  meteofrance_arpege_europe: 'ARPEGE · MeteoFrance',
  jma_msm: 'JMA MSM',
  gem_seamless: 'GEM · Canada',
}

export interface HourlyPoint {
  time: number
  temperature?: number
  precipitation?: number
  precipitationProbability?: number
  windSpeed?: number
  windGust?: number
  windDirection?: number
  cloudCover?: number
  weatherCode?: number
  humidity?: number
  apparentTemperature?: number
  dewPoint?: number
  pressure?: number
  visibility?: number
  uvIndex?: number
  soilMoisture0_7?: number
  soilTemperature0_7?: number
  evapotranspiration?: number
  snowfall?: number
}

export interface ModelForecast {
  model: WeatherModelId
  hourly: HourlyPoint[]
}

export type DisagreementLevel = 'low' | 'medium' | 'high'

export interface FusedScalar {
  mean: number
  stdDev: number
  lower: number
  upper: number
  disagreement: DisagreementLevel
  perModel: Array<{ model: WeatherModelId; value: number }>
}

export interface FusedPrecip {
  meanMM: number
  probability: number
  lowMM: number
  highMM: number
  modelsExpectingRain: number
  totalModels: number
  perModel: Array<{ model: WeatherModelId; value: number }>
}

export interface FusedHour {
  time: number
  temperature?: FusedScalar
  apparentTemperature?: FusedScalar
  precipitation?: FusedPrecip
  windSpeed?: FusedScalar
  windGust?: FusedScalar
  windDirection?: FusedScalar
  cloudCover?: FusedScalar
  weatherCodeMode?: number
  humidity?: FusedScalar
  dewPoint?: FusedScalar
  pressure?: FusedScalar
  visibility?: FusedScalar
  uvIndex?: FusedScalar
  soilMoisture?: FusedScalar
  soilTemperature?: FusedScalar
  evapotranspiration?: FusedScalar
  snowfall?: FusedScalar
  contributingModels: WeatherModelId[]
}

export interface Coordinate {
  latitude: number
  longitude: number
}

File: ios/Wetterfusion/Sources/Models/Forecast.swift
import Foundation

enum WeatherModel: String, CaseIterable, Codable, Identifiable {
    case iconD2 = "icon_d2"
    case iconEu = "icon_eu"
    case iconGlobal = "icon_global"
    case ecmwfIfs = "ecmwf_ifs025"
    case ecmwfAifs = "ecmwf_aifs025"
    case gfs = "gfs_seamless"
    case meteoFranceArome = "meteofrance_arome_france_hd"
    case meteoFranceArpege = "meteofrance_arpege_europe"
    case jmaMsm = "jma_msm"
    case gem = "gem_seamless"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .iconD2: return "ICON-D2 (DWD)"
        case .iconEu: return "ICON-EU (DWD)"
        case .iconGlobal: return "ICON-Global (DWD)"
        case .ecmwfIfs: return "ECMWF IFS"
        case .ecmwfAifs: return "ECMWF AIFS"
        case .gfs: return "GFS (NOAA)"
        case .meteoFranceArome: return "AROME (MeteoFrance)"
        case .meteoFranceArpege: return "ARPEGE (MeteoFrance)"
        case .jmaMsm: return "JMA MSM"
        case .gem: return "GEM (Canada)"
        }
    }
}

struct HourlyForecast: Identifiable, Codable {
    let id = UUID()
    let time: Date
    let temperature: Double?
    let precipitation: Double?
    let precipitationProbability: Double?
    let windSpeed: Double?
    let windGust: Double?
    let cloudCover: Double?
    let weatherCode: Int?

    enum CodingKeys: String, CodingKey {
        case time, temperature, precipitation, precipitationProbability, windSpeed, windGust, cloudCover, weatherCode
    }
}

struct ModelForecast: Identifiable {
    let id = UUID()
    let model: WeatherModel
    let hourly: [HourlyForecast]
}

struct FusedHourlyPoint: Identifiable {
    let id = UUID()
    let time: Date

    let temperature: FusedValue?
    let precipitation: FusedPrecip?
    let windSpeed: FusedValue?

    let contributingModels: [WeatherModel]
}

struct FusedValue {
    let mean: Double
    let stdDev: Double
    let lower: Double
    let upper: Double

    var disagreement: Disagreement {
        let coeff = abs(mean) > 0.1 ? stdDev / abs(mean) : stdDev
        if coeff < 0.05 { return .low }
        if coeff < 0.15 { return .medium }
        return .high
    }
}

struct FusedPrecip {
    let meanMM: Double
    let probability: Double
    let lowMM: Double
    let highMM: Double
    let modelsExpectingRain: Int
    let totalModels: Int
}

enum Disagreement {
    case low, medium, high

    var label: String {
        switch self {
        case .low: return "Modelle einig"
        case .medium: return "leichte Unsicherheit"
        case .high: return "Modelle uneinig"
        }
    }
}
